File: python/kmerfreq.py
"""
Determine the k-mer frequencies in a FASTA file. 
The value of K should be an argument with a default parameter (e.g. 3). 
Output format should include tab-separated and JSON.

Inputs

Multi-FASTA file (gzipped or STDIN)
K-mer size

Outputs

TSV
JSON
"""

import argparse
import csv
import json
from readfasta import read_record

parser = argparse.ArgumentParser(description='Determine the kmer frequences \
in a fasta file')
parser.add_argument('fasta', type = str, help = 'path to file')
parser.add_argument('-K', type = int, required = False, help = 'K-mer size', default = 3)
parser.add_argument('-j', action = "store_true", help = " ")
arg = parser.parse_args()

total = 0
freq = {}
for id, seq in read_record(arg.fasta):
	for i in range(len(seq) - arg.K + 1):
		kmer = seq[i:i+arg.K]
		if kmer not in freq: freq[kmer] = 0
		freq[kmer] += 1
		total += 1
		
for kmer in freq: freq[kmer] /= total
if arg.j:
	print(json.dumps(freq, indent=4))
else:
	for key in freq:
		print(f'{key}	{freq[key]}')

File: alan/go/kmerfreq/kmerfreq.go
package main

import (
	"github.com/AlanAloha/read_fasta"
	"flag"
	"sort"
	"fmt"
	"os"
)

func main() {
	fasta := flag.String("in", "", "path to fasta file (required)")
	k := flag.Int("k", 3, "k-mer size")
	j := flag.Bool("j", false, "output in json (default: tab-separated)")
	flag.Parse()
	
	if *fasta == "" {
		flag.Usage()
		os.Exit(1)
	}
	
	fmt.Println(*fasta, *k, *j)
	
	total := 0.0
	freq := make(map[string]float64)
	
	var kmer string
	read_fasta.Read_fasta(*fasta, func (read read_fasta.Read) {
		for i := 0; i < (len(read.Seq) - *k + 1); i++ {
			kmer = read.Seq[i:i+*k]
			freq[kmer] += 1
			total += 1
		}
	})
	
	for kmer := range freq {freq[kmer] /= total}
	
	//create a slice to sort the keys
	kmers := make([]string, 0)
	for kmer := range freq {kmers = append(kmers, kmer)}
	sort.Strings(kmers)
	
	
	if *j {
		fmt.Printf("{\n")
		for i, kmer := range kmers {
			if i == len(kmers) - 1 {
				fmt.Printf("\t\"%s\": %f\n", kmer, freq[kmer])
			} else {
				fmt.Printf("\t\"%s\": %f,\n", kmer, freq[kmer])
			}
		}
		fmt.Printf("}\n")
	} else {
		for _, kmer := range kmers {
			fmt.Printf("%s\t%f\n", kmer, freq[kmer])
		}
	}

	
	
	

}
